How-to-upgrade-review-version.md

```diff
@@ -0,0 +1,399 @@
+where you are overwhelmed by the amount of work you need to do and
+just want a way to reduce it for a while. You can even put multiple
+`declare module`s in a single file so that all your workarounds are in
+one place.
+
+This fixes the immediate problem, but a number of errors pops up on
+`PatternVisitor extends OriginalPatternVisitor`. So the definition
+needs to be more complete. Here's what I ended up with:
+
+```ts
+declare module "eslint-scope/lib/pattern-visitor" {
+    import { Node, Expression, SpreadElement } from "estree";
+    type Options = {
+        topLevel: boolean;
+        rest: boolean;
+        assignments: any[];
+    };
+    class PatternVisitor {
+        constructor(
+            options: any,
+            rootPattern: any,
+            callback: (p: any, options: Options) => void);
+        rightHandNodes: Array<Expression | SpreadElement>;
+
+        Identifier(node: Node): void;
+        visit(node: Node): void;
+    }
+    export = PatternVisitor;
+}
+```
+
+I got this by (1) looking at pattern-visitor.js and (2) the estree
+types. eslint-scope uses estree types without explicitly referring to
+them, so an eventual correct fix would change eslint-scope. For now, I
+have enough to keep going.
+
+## Fix errors in existing types
+
+Unfortunately, the improved type for pattern-visitor once again causes
+an error on `new PatternVisitor`. This time, the callback's type,
+`Function` isn't specific enough to work with the specific function
+type of the callback:
+
+``ts
+callback: (p: any, options: Options) => void
+```
+
+So the existing JSDoc type annotation needs to change:
+
+```js
+    /**
+     * Override to use PatternVisitor we overrode.
+     * @param {Identifier} node The Identifier node to visit.
+     * @param {Object} [options] The flag to visit right-hand side nodes.
+     * @param {Function} callback The callback function for left-hand side nodes.
+     * @returns {void}
+     */
+    visitPattern(node, options, callback) {
+```
+
+I changed `@param {Function} callback` to the more precise
+`@param {(p: any, options: Options) => void} callback`.
+
+Except that Options isn't exported by my pattern-visitor workaround.
+So I moved it in the global namespace, outside the `declare module`:
+
+```ts
+type Options = {
+    topLevel: boolean;
+    rest: boolean;
+    assignments: any[];
+};
+
+declare module "eslint-scope/lib/pattern-visitor" {
+    import { Node, Expression, SpreadElement } from "estree";
+    class PatternVisitor {
+        constructor(
+            options: any,
+            rootPattern: any,
+            callback: (p: any, options: Options) => void);
+        rightHandNodes: Array<Expression | SpreadElement>;
+
+        Identifier(node: Node): void;
+        visit(node: Node): void;
+    }
+    export = PatternVisitor;
+}
+```
+
+This is messy but at the same time *horribly* convenient.
+
+## Add type annotations to everything else
```

Comment:
I'm probably wrong, but "type annotation" to me means type annotations in source like in TypeScript and we can't actually do that here. Should you clarify that you're talking about JSDoc?

Reply:
I'll clarify it. I work so much with JSDoc I forget that it's not the default.